### ISSUE TITLE
0.3.2 janusgraph gremlin step "or()" has issue

### ISSUE THREAD

**Issue by lucky-yang** [JanusGraph/janusgraph#1714]
0.3.2 janusgraph gremlin step "or()" has the issue as below

after we selected some vertices used by step "or()", and continue to use step "as() & select()", it will not return the result. 

Please refer to the below sql step result:

gremlin> g.V().has("vertexLabel","category").or(__.has("rootSchemaLogicKey","service").has("schemaLogicKey", "14702"),__.has("rootSchemaLogicKey","service").has("schemaLogicKey", "17136"),__.has("rootSchemaLogicKey","service").has("schemaLogicKey", "17040")).as("aa")
==>v[11509952]
==>v[21385416]
==>v[21397704]

gremlin> g.V().has("vertexLabel","category").or(__.has("rootSchemaLogicKey","service").has("schemaLogicKey", "14702"),__.has("rootSchemaLogicKey","service").has("schemaLogicKey", "17136"),__.has("rootSchemaLogicKey","service").has("schemaLogicKey", "17040")).as("aa").select("aa")
gremlin>

**Comment by vtslab** [comment]
Confirmed, good catch! See the transcript below for version 0.3.1/0.3.2 with the Graph of the Gods:
```
gremlin> graph = JanusGraphFactory.open("inmemory")
==>standardjanusgraph[inmemory:[127.0.0.1]]
gremlin> GraphOfTheGodsFactory.loadWithoutMixedIndex(graph,true)
==>null
gremlin> g = graph.traversal()
==>graphtraversalsource[standardjanusgraph[inmemory:[127.0.0.1]], standard]

gremlin> g.V().or(__.has('name', 'pluto').has('age', 4000), __.has('name', 'hercules').has('age', 30)).as('x')
==>v[16448]
==>v[8256]

gremlin> g.V().or(__.has('name', 'pluto').has('age', 4000), __.has('name', 'hercules').has('age', 30)).as('x').select('x')  // !!!

gremlin> g.V().or(__.has('name', 'pluto'), __.has('name', 'hercules')).as('x').select('x').valueMap(true)
==>[id:16448,name:[pluto],label:god,age:[4000]]
==>[id:8256,name:[hercules],label:demigod,age:[30]]
```
In JanusGraph-0.2.0 the line commented with !!! returns two vertices, as expected.

**Comment by vtslab** [comment]
The issue is already visible from the traversal explanations as the @[x] reference disappears somehow in the misbehaving query.

The good one as reference (v0.3.2):
```
gremlin> g.V().or(__.has('name', 'pluto'), __.has('name', 'hercules')).as('x').select('x').explain()
==>Traversal Explanation
================================================================================
Original Traversal                          [GraphStep(vertex,[]), OrStep([[H
                                               asStep([name.eq(pluto)])], [HasSt
                                               ep([name.eq(hercules)])]])@[x], S
                                               electOneStep(last,x)]

ConnectiveStrategy                    [D]   [GraphStep(vertex,[]), OrStep([[H
                                               asStep([name.eq(pluto)])], [HasSt
                                               ep([name.eq(hercules)])]])@[x], S
                                               electOneStep(last,x)]
MatchPredicateStrategy                [O]   [GraphStep(vertex,[]), OrStep([[H
                                               asStep([name.eq(pluto)])], [HasSt
                                               ep([name.eq(hercules)])]])@[x], S
                                               electOneStep(last,x)]
FilterRankingStrategy                 [O]   [GraphStep(vertex,[]), OrStep([[H
                                               asStep([name.eq(pluto)])], [HasSt
                                               ep([name.eq(hercules)])]])@[x], S
                                               electOneStep(last,x)]
InlineFilterStrategy                  [O]   [GraphStep(vertex,[]), HasStep([n
                                               ame.or(eq(pluto), eq(hercules))])
                                               @[x], SelectOneStep(last,x)]
IncidentToAdjacentStrategy            [O]   [GraphStep(vertex,[]), HasStep([n
                                               ame.or(eq(pluto), eq(hercules))])
                                               @[x], SelectOneStep(last,x)]
AdjacentToIncidentStrategy            [O]   [GraphStep(vertex,[]), HasStep([n
                                               ame.or(eq(pluto), eq(hercules))])
                                               @[x], SelectOneStep(last,x)]
RepeatUnrollStrategy                  [O]   [GraphStep(vertex,[]), HasStep([n
                                               ame.or(eq(pluto), eq(hercules))])
                                               @[x], SelectOneStep(last,x)]
PathRetractionStrategy                [O]   [GraphStep(vertex,[]), HasStep([n
                                               ame.or(eq(pluto), eq(hercules))])
                                               @[x], SelectOneStep(last,x)]
CountStrategy                         [O]   [GraphStep(vertex,[]), HasStep([n
                                               ame.or(eq(pluto), eq(hercules))])
                                               @[x], SelectOneStep(last,x)]
LazyBarrierStrategy                   [O]   [GraphStep(vertex,[]), HasStep([n
                                               ame.or(eq(pluto), eq(hercules))])
                                               @[x], SelectOneStep(last,x)]
AdjacentVertexFilterOptimizerStrategy [P]   [GraphStep(vertex,[]), HasStep([n
                                               ame.or(eq(pluto), eq(hercules))])
                                               @[x], SelectOneStep(last,x)]
JanusGraphLocalQueryOptimizerStrategy [P]   [GraphStep(vertex,[]), HasStep([n
                                               ame.or(eq(pluto), eq(hercules))])
                                               @[x], SelectOneStep(last,x)]
JanusGraphStepStrategy                [P]   [JanusGraphStep([],[name.or(=(plu
                                               to), =(hercules))])@[x], SelectOn
                                               eStep(last,x)]
ProfileStrategy                       [F]   [JanusGraphStep([],[name.or(=(plu
                                               to), =(hercules))])@[x], SelectOn
                                               eStep(last,x)]
StandardVerificationStrategy          [V]   [JanusGraphStep([],[name.or(=(plu
                                               to), =(hercules))])@[x], SelectOn
                                               eStep(last,x)]

Final Traversal                             [JanusGraphStep([],[name.or(=(plu
                                               to), =(hercules))])@[x], SelectOn
                                               eStep(last,x)]

```
The misbehaving one (v0.3.2):
```
gremlin> g.V().or(__.has('name', 'pluto').has('age', 4000), __.has('name', 'hercules').has('age', 30)).as('x').select('x').explain()
==>Traversal Explanation
================================================================================
Original Traversal                          [GraphStep(vertex,[]), OrStep([[H
                                               asStep([name.eq(pluto), age.eq(40
                                               00)])], [HasStep([name.eq(hercule
                                               s), age.eq(30)])]])@[x], SelectOn
                                               eStep(last,x)]

ConnectiveStrategy                    [D]   [GraphStep(vertex,[]), OrStep([[H
                                               asStep([name.eq(pluto), age.eq(40
                                               00)])], [HasStep([name.eq(hercule
                                               s), age.eq(30)])]])@[x], SelectOn
                                               eStep(last,x)]
MatchPredicateStrategy                [O]   [GraphStep(vertex,[]), OrStep([[H
                                               asStep([name.eq(pluto), age.eq(40
                                               00)])], [HasStep([name.eq(hercule
                                               s), age.eq(30)])]])@[x], SelectOn
                                               eStep(last,x)]
FilterRankingStrategy                 [O]   [GraphStep(vertex,[]), OrStep([[H
                                               asStep([name.eq(pluto), age.eq(40
                                               00)])], [HasStep([name.eq(hercule
                                               s), age.eq(30)])]])@[x], SelectOn
                                               eStep(last,x)]
InlineFilterStrategy                  [O]   [GraphStep(vertex,[]), OrStep([[H
                                               asStep([name.eq(pluto), age.eq(40
                                               00)])], [HasStep([name.eq(hercule
                                               s), age.eq(30)])]])@[x], SelectOn
                                               eStep(last,x)]
IncidentToAdjacentStrategy            [O]   [GraphStep(vertex,[]), OrStep([[H
                                               asStep([name.eq(pluto), age.eq(40
                                               00)])], [HasStep([name.eq(hercule
                                               s), age.eq(30)])]])@[x], SelectOn
                                               eStep(last,x)]
AdjacentToIncidentStrategy            [O]   [GraphStep(vertex,[]), OrStep([[H
                                               asStep([name.eq(pluto), age.eq(40
                                               00)])], [HasStep([name.eq(hercule
                                               s), age.eq(30)])]])@[x], SelectOn
                                               eStep(last,x)]
RepeatUnrollStrategy                  [O]   [GraphStep(vertex,[]), OrStep([[H
                                               asStep([name.eq(pluto), age.eq(40
                                               00)])], [HasStep([name.eq(hercule
                                               s), age.eq(30)])]])@[x], SelectOn
                                               eStep(last,x)]
PathRetractionStrategy                [O]   [GraphStep(vertex,[]), OrStep([[H
                                               asStep([name.eq(pluto), age.eq(40
                                               00)])], [HasStep([name.eq(hercule
                                               s), age.eq(30)])]])@[x], SelectOn
                                               eStep(last,x)]
CountStrategy                         [O]   [GraphStep(vertex,[]), OrStep([[H
                                               asStep([name.eq(pluto), age.eq(40
                                               00)])], [HasStep([name.eq(hercule
                                               s), age.eq(30)])]])@[x], SelectOn
                                               eStep(last,x)]
LazyBarrierStrategy                   [O]   [GraphStep(vertex,[]), OrStep([[H
                                               asStep([name.eq(pluto), age.eq(40
                                               00)])], [HasStep([name.eq(hercule
                                               s), age.eq(30)])]])@[x], SelectOn
                                               eStep(last,x)]
AdjacentVertexFilterOptimizerStrategy [P]   [GraphStep(vertex,[]), OrStep([[H
                                               asStep([name.eq(pluto), age.eq(40
                                               00)])], [HasStep([name.eq(hercule
                                               s), age.eq(30)])]])@[x], SelectOn
                                               eStep(last,x)]
JanusGraphLocalQueryOptimizerStrategy [P]   [GraphStep(vertex,[]), OrStep([[H
                                               asStep([name.eq(pluto), age.eq(40
                                               00)])], [HasStep([name.eq(hercule
                                               s), age.eq(30)])]])@[x], SelectOn
                                               eStep(last,x)]
JanusGraphStepStrategy                [P]   [Or(JanusGraphStep([],[name.eq(pl
                                               uto), age.eq(4000)]),JanusGraphSt
                                               ep([],[name.eq(hercules), age.eq(
                                               30)])), SelectOneStep(last,x)]
ProfileStrategy                       [F]   [Or(JanusGraphStep([],[name.eq(pl
                                               uto), age.eq(4000)]),JanusGraphSt
                                               ep([],[name.eq(hercules), age.eq(
                                               30)])), SelectOneStep(last,x)]
StandardVerificationStrategy          [V]   [Or(JanusGraphStep([],[name.eq(pl
                                               uto), age.eq(4000)]),JanusGraphSt
                                               ep([],[name.eq(hercules), age.eq(
                                               30)])), SelectOneStep(last,x)]

Final Traversal                             [Or(JanusGraphStep([],[name.eq(pl
                                               uto), age.eq(4000)]),JanusGraphSt
                                               ep([],[name.eq(hercules), age.eq(
                                               30)])), SelectOneStep(last,x)]
```

**Comment by caamartin35** [comment]
The workaround mentioned by @vtslab is works functionally. However, if you are using indexes it will fail to use them. I find the following to be a suitable workaround after some testing:

`
gremlin> g.V().or(has('name', 'pluto').has('age', 4000), __.has('name', 'hercules').has('age', 30)).identity().as('x').select('x').valueMap()
==>[name:[hercules],age:[30]]
==>[name:[pluto],age:[4000]]
`

**Comment by mad** [comment]
Fixed #1943
